### ISSUE TITLE
Build system

### ISSUE THREAD

**Issue by defanor** [endless-sky/endless-sky#1882]
It would be nice (at least for some users and systems) to introduce
the following build-related changes:

- Provide an `INSTALL` file, which would include instructions along
  with a list of dependencies.

- Use, at least as an alternative, something more common than `scons`:
  if not GNU build system, at least a `Makefile`, or a shell
  script. While the game itself doesn't seem to depend on Python,
  scons (which also requires Python 2, not 3) pulls dependencies that
  would be unnecessary otherwise.

**Comment by vflyson** [comment]
I second this proposal. It seems to be a very good open-source game, yet it's a shame that it's not that easy to stay up-to-date with the latest version if you're using an open-source OS.

**Comment by defanor** [comment]
Nice. I think it is an improvement, though there is a couple of small
issues with that, too:

- It requires cmake 3.1.0 or newer there, while Debian 8 (jessie,
  current stable) only has 3.0.2 in its repositories, and CentOS/RHEL 7
  has 2.8.12.

- Cmake itself is not that minimalistic, and IME it's not installed on a
  system more often than it is -- so, still additional dependencies just
  to build the game.

**Comment by lheckemann** [comment]
Yeah, tbh I think it could probably be done with a plain makefile and I'll have a go at doing so tomorrow. Happy new year!

**Comment by defanor** [comment]
Shouldn't be hard to compose a basic Makefile, indeed, since it can be
built even with just `g++ -std=c++11 -Wall source/*.cpp -lSDL2 -lpng
-ljpeg -lGL -lGLEW -lopenal -lmad`, though then there's a lot of special
cases to cover, like installation prefix, additional options, clean
uninstall, etc.

Meantime, I've tried to build it with autotools, but to make it work
nicely, it would require to adjust the sources as well -- to use correct
paths and such (looks like they are currently hardcoded in `Files.cpp`,
even though the `SConstruct` file appears to handle custom
prefixes). It's also quite cumbersome to install all the data files, so
not sure if I will continue; in case if somebody else would want to do
that, here is what I've got:
<https://gist.github.com/defanor/b9adc5fafbb6ee28765d0f3a978b424d>.

And happy new year to you too.

**Comment by jraynal** [comment]
Hi, I do realize I am late to the party, I was curious about CMake support. I am using CMake quite extensively in my day-job, and I think it could bring some interesting aspects that a simple Makefile can't do. (I never designed build systems with autotools).

As for CMake version, the higher the better. It is really easy to download the latest CMake build, and install it locally. The more recent versions have better support for mobile development, and can generate project files for more recent IDEs (QtCreator, VS, ...).

More importantly, with the `CPack` tools, it becomes really easy to package the game for Ubuntu/Linux/OSX! Which is mentioned in the [help wanted](https://github.com/endless-sky/endless-sky/wiki/HelpWanted#programming-tasks) section of the wiki.

I would be happy to help reviving the [cmake](https://github.com/lheckemann/endless-sky/tree/cmake) branch, as I spotted some potential improvements.

Sorry for rambling... :slightly_smiling_face:

**Comment by KiLEdEnNis** [comment]
no need for a sorry, as long as an issue is open and your reaction is relevant you're free to comment

**Comment by lheckemann** [comment]
@jraynal yeah, that would be cool! I think it shouldn't be too difficult to update my branch, let me know if you have any difficulties.

**Comment by endless-sky** [comment]
In general, I want to limit how many different build scripts need to be updated every time a new source code file gets added. The existing scons script works fine for Linux (which is where I do all the development work) and has the advantage that the file doesn't have to be changed every time source code is added (because it just compiles everything in the "source" folder).

**Comment by jraynal** [comment]
I can hear that ;), maintaining build scripts is a pain. Note that we can keep the file globing advantage, and limit the maintenance needed in CMake. What I was thinking was more in term of packaging advantages, and then scalability. It might be completely irrelevant at this point, but CMake's CPack can generate installers for Linux, Windows and MacOSX. I was also able to generate a XCode project from the CMake scripts, which would remove the need of maintaining the currently available xcodeproject, as well as centralize dependency resolution... I haven't been able to reproduce all the functionality of the current build system just now though.

**Comment by jraynal** [comment]
I haven't touched the build system since my last post above... I had some troubles with the MacOSX installation. Finding the dependencies and automatically pulling the sources was not a problem at all. As for the version, anything post 3.2 is perfectly fine for most Unix systems. 3.7.2 is what I use for Android (they introduced CMake server in 3.7 so that it would play nicely with IDEs). That being said, using an early version probably offers better compatibility with VS.

I kind of stopped because last year it didn't seem like a high priority, but I could get back to it if it would help :).

**Comment by OrdoFlammae** [comment]
I can work with CMake, and I think there's a possibility that we can use CPack to create the installers (I don't know if SCons has a similar function, or how the Windows binary downloads are made), so that should be able to take care of all the requirements we have. However, I have never even looked at CPack before now, so that may not do what we need it to do.

**Comment by jraynal** [comment]
The way I usually present it is that CPack just wraps the installation step in a portable installation package. Last december, I worked on [this CMakeLists.txt](https://github.com/jraynal/endless-sky/blob/cmake/CMakeLists.txt), and as I mentioned, installing on OSX gave me some troubles.

I never worked with windows, it seems that this project has the extra challenge of requiring Steam compatibility which is completely unknown to me!

**Comment by OrdoFlammae** [comment]
Really? I never touched Steam, and yet I still built it on Ubuntu. Why would that be?

**Comment by jraynal** [comment]
Just gut feeling. Steam bundles Windows, MacOSX and Linux variants. They seem use their own installer, they probably enforce a specific directory structure. CMake can do single configuration builds really well, in my experience it falls short when you need to generate a cross-platform package, but that might be my own aversion of CTest. In my company we have a bash script to generate all the variants we want for our library.

The second worry I have is how to iterate and test that the installation generated is correct. I can only assume Steam offers all the tools needed.

That being said, there is only a few ways to know ;).

**Comment by jafdy** [comment]
Steam doesn't "install" ES as such, it downloads all the files and places them in a directory together in much the same way you would when building it yourself. The main difference is that steam uses a pre-built executable.

**Comment by TomGoodIdea** [comment]
Is this issue still relevant if we support CMake now?

**Comment by tibetiroka** [comment]
We still support SCons, though I'm not sure what dependencies it has today.

I suppose the same arguments could be made about CMake's dependencies, though. It might pull stuff in that we don't need, such as `curl`.

**Comment by TheGiraffe3** [comment]
Since we no longer support SCons, is this issue still relevent?

**Comment by defanor** [comment]
> Since we no longer support SCons, is this issue still relevent?

Perhaps not, at least I would be fine with it being closed: there are
build instructions now, and though the CMake version requirement went up
in those, so did the CMake versions in stable distributions'
repositories (particularly Debian 12 has version 3.25 now, while build
instructions require 3.16 and recommend 3.21). Apparently ninja is also
used now, which depends on Python, like SCons did, but at least on
Python 3. While a lower-dependency build is possible, the current one
looks fine to me.
